File: Solutions/Python/minimum_genetic_mutation.py
"""
    [MEDIUM]
    433. Minimum Genetic Mutation

    Concepts:
    - string
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 17.96 MB  [Beats 19.13%]
"""

from collections import deque

class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        """
            BFS

            Explore every subtree until either leaf reached
            (string not in bank) or end reached

            If end not reached, return -1
        """
        bank = set(bank + [start])
        count = 0

        def get_distance(s1, s2):
            return len([i for i in range(len(s1)) if s1[i] != s2[i]])

        q = deque([(start, 0)])
        while q:
            len_q = len(q)

            for _ in range(len_q):
                curr, mutations = q.popleft()

                if curr not in bank:
                    continue
                bank.remove(curr)

                if curr == end:
                    return mutations

                for string in bank:
                    if get_distance(string, curr) == 1:
                        q.append((string, mutations + 1))
                
        return -1

File: Solutions/Cpp/minimum_genetic_mutation.cpp
/*
    [MEDIUM]
    433. Minimum Genetic Mutation

    Concepts:
    - string
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 8.93 MB   [Beats 28.97%]
*/

class Solution {
public:
    int get_distance(std::string s1, std::string s2) {
        int diff_count = 0;
        for (int i = 0; i < s1.size(); i++)
            if (s1[i] != s2[i])
                diff_count++;
        return diff_count;
    }
    
    int minMutation(string start, string end, vector<string>& b) {
        /*
            BFS

            Explore every subtree until either leaf reached
            (string not in bank) or end reached

            If end not reached, return -1
        */
        std::unordered_set<std::string> bank(b.begin(), b.end());
        bank.insert(start);
        int count = 0;

        std::queue<std::pair<std::string, int>> q;
        q.push(std::make_pair(start, 0));
        int len_q, mutations;
        std::string curr;

        while (!q.empty()) {
            len_q = q.size();

            for (int _ = 0; _ < len_q; _++) {
                auto [curr, mutations] = q.front();
                q.pop();

                if (!bank.contains(curr))
                    continue;
                bank.erase(curr);

                if (curr == end)
                    return mutations;

                for (std::string str : bank)
                    if (get_distance(str, curr) == 1)
                        q.push({str, mutations + 1});
            }
        }

        return -1;
    }
};